src/java/org/apache/cassandra/transport/Message.java

```diff
@@ -225,12 +237,31 @@ protected boolean isTraceable()
         /**
          * Returns the time elapsed since this request was created. Note that this is the total lifetime of the request
          * in the system, so we expect increasing returned values across multiple calls to elapsedTimeSinceCreation.
+         * The elaspse time is calculated using the start time from the custom payload if it exists, otherwise it falls
+         * back to the time the request was created in the NTR stage (that is, when it was deserialized).
          *
          * @param timeUnit the time unit in which to return the elapsed time
          * @return the time elapsed since this request was created
          */
         protected long elapsedTimeSinceCreation(TimeUnit timeUnit)
         {
+            Map<String, ByteBuffer> customPayload = getCustomPayload();
+            if (customPayload != null && customPayload.containsKey(CNDB_START_TIME))
+            {
+                ByteBuffer startTime = customPayload.get(CNDB_START_TIME);
+                try
+                {
+                    long startTimeMillis = ByteBufferUtil.toLong(startTime);
+                    long startTimeNanos = TimeUnit.MILLISECONDS.toNanos(startTimeMillis);
```

Comment:
I suggest we change the resolution (and name of the CNDB_START_TIME) key (will coordinate the change in CQL router) to use Nanos and spare the conversions (generally timeout in C* are Nano based).

The name also doesn't need a CNDB prefix. Also START_TIME maybe confused with when the request was originally created (i.e. client side). So all things considered, I'd suggest something like:
CQL_CREATION_TIME_NANOS

Reply:
Changed to `REQUEST_CREATE_NANOS`